mon: revamp health check/warning system

liewegas opened the pull request:
Implements new health checks with more structure.  Old checks are left untouched but will only be used for the duration of the upgrade to luminous; once the upgrade completes the new checks are used instead.

Caveats
- Monitor::log_health should atomically include the generated log msgs in the commit.
- It is somewhat less convenient to log problems with the quorum itself in the official health history.  This is partly because before it was a total mess and now we have a structure, so that's not necessarily a bad ting.  Mostly it  highlights the fact that we're tying health alerts explicitly to paxos consensus state.
- the format of 'ceph health detail' (plaintext) is different: we indent to group details by summary.  If you need something that is compatible with pre-luminous you can enable mon_health_preluminous_compat=true and the json/xml output will include both smooshed together (plaintext will look roughly like the old format).
- the structured format of 'ceph health [detail] -f wahtever' has lost the extra weird crap about timechecks etc that didn't look like other health checks.

TODO
- [x] OSDMonitor checks
- [x] PGMap to generate a health_check_map_t
- [x] update mgr -> mon messaging and MgrStatMonitor health structs to be health_check_map_t
- [x] MDSMonitor new-style checks

That's about it!

qa
- [x] rados
- [x] rgw
- [x] rbd
- [x] fs

tchaikov commented on src/mon/HealthMonitor.cc:
```diff
@@ -0,0 +1,286 @@
+#include <stdlib.h>
+#include <limits.h>
+#include <sstream>
+#include <regex>
```
please see http://tracker.ceph.com/issues/20112

wjwithagen replied:
@tchaikov @liewegas 
I would also be worried about differences in regex implementations between Linux and BSDs. I've already run into some. Chances are that Boost::Regex is going to be a more stabil denominator.

tchaikov commented on src/mon/health_check.h:
```diff
@@ -0,0 +1,147 @@
+
+  friend bool operator==(const health_check_map_t& l,
+			 const health_check_map_t& r) {
+    if (l.checks.size() != r.checks.size()) {
```
`std::map` offers `operator==()`, which does exactly the same thing.

tchaikov commented on src/mon/health_check.h:
```diff
@@ -0,0 +1,147 @@
+    }
+    return true;
+  }
+  friend bool operator!=(const health_check_map_t& l,
```
and also `operator!=`.

dmick commented:
You probably are, but, be aware that deployment tools, custom monitoring scripts, health checks, etc. are all going to be impacted, particularly if you change the formatted output (which is assumed to be more stable).

knandya commented:
I agree with @dmick opinion on this. Consumers of current health checks made custom changes for their monitoring (Sensu, Nagios etc) and created their own alerts.

liewegas commented:
There is a middle ground.. the json fields of the old format and the new format are mostly non-overlapping, so we could include them both for the next cycle.  The overlap currently is that the 'detail' section previous was an array of strings and now it's a dict of error codes to arrays.  We could rename it, or (perhaps better) move the 'detail' list for each health alert as an item in that dict instead of a separate top-level structure.  (This is better for programmatic parsing IMO, but a bit less friendly for humans reading the JSON.)

leseb commented:
@liewegas can we get a sample output once you're done?
Thanks!

liewegas commented:
Before:
<pre>
{
    "summary": [
        {
            "severity": "HEALTH_WARN",
            "summary": "8 pgs degraded"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "8 pgs stale"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "8 pgs undersized"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "1 root (2 osds) down"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "1 host (2 osds) down"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "1 osds down"
        }
    ],
    "overall_status": "HEALTH_WARN",
    "detail": [ 
        "pg 0.7 is stale+active+undersized+degraded, acting [0]",
        "pg 0.6 is stale+active+undersized+degraded, acting [0]",
        "pg 0.5 is stale+active+undersized+degraded, acting [0]",
        "pg 0.4 is stale+active+undersized+degraded, acting [0]",
        "pg 0.0 is stale+active+undersized+degraded, acting [0]",
        "pg 0.1 is stale+active+undersized+degraded, acting [0]",
        "pg 0.2 is stale+active+undersized+degraded, acting [0]",
        "pg 0.3 is stale+active+undersized+degraded, acting [0]",
        "root default (2 osds) is down",
        "host gnit (root=default) (2 osds) is down",
        "osd.0 (root=default,host=gnit) is down"
    ]
}
HEALTH_WARN 8 pgs degraded; 8 pgs stale; 8 pgs undersized; 1 root (2 osds) down; 1 host (2 osds) down; 1 osds down
pg 0.7 is stale+active+undersized+degraded, acting [0]
pg 0.6 is stale+active+undersized+degraded, acting [0]
pg 0.5 is stale+active+undersized+degraded, acting [0]
pg 0.4 is stale+active+undersized+degraded, acting [0]
pg 0.0 is stale+active+undersized+degraded, acting [0]
pg 0.1 is stale+active+undersized+degraded, acting [0]
pg 0.2 is stale+active+undersized+degraded, acting [0]
pg 0.3 is stale+active+undersized+degraded, acting [0]
root default (2 osds) is down
host gnit (root=default) (2 osds) is down
osd.0 (root=default,host=gnit) is down
</pre>
after 
<prE>
{
    "checks": {
        "PG_DEGRADED": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs degraded"
        },
        "PG_STALE": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs stale"
        },
        "PG_UNDERSIZED": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs undersized"
        }
    },
    "status": "HEALTH_WARN",
    "detail": { 
        "PG_DEGRADED": [
            "pg 0.7 is stale+active+undersized+degraded, acting [0]",
            "pg 0.6 is stale+active+undersized+degraded, acting [0]",
            "pg 0.5 is stale+active+undersized+degraded, acting [0]",
            "pg 0.4 is stale+active+undersized+degraded, acting [0]",
            "pg 0.0 is stale+active+undersized+degraded, acting [0]",
            "pg 0.1 is stale+active+undersized+degraded, acting [0]",
            "pg 0.2 is stale+active+undersized+degraded, acting [0]",
            "pg 0.3 is stale+active+undersized+degraded, acting [0]"
        ],
        "PG_STALE": [
            "pg 0.7 is stale+active+undersized+degraded, acting [0]",
            "pg 0.6 is stale+active+undersized+degraded, acting [0]",
            "pg 0.5 is stale+active+undersized+degraded, acting [0]",
            "pg 0.4 is stale+active+undersized+degraded, acting [0]",
            "pg 0.0 is stale+active+undersized+degraded, acting [0]",
            "pg 0.1 is stale+active+undersized+degraded, acting [0]",
            "pg 0.2 is stale+active+undersized+degraded, acting [0]",
            "pg 0.3 is stale+active+undersized+degraded, acting [0]"
        ],
        "PG_UNDERSIZED": [
            "pg 0.7 is stale+active+undersized+degraded, acting [0]",
            "pg 0.6 is stale+active+undersized+degraded, acting [0]",
            "pg 0.5 is stale+active+undersized+degraded, acting [0]",
            "pg 0.4 is stale+active+undersized+degraded, acting [0]",
            "pg 0.0 is stale+active+undersized+degraded, acting [0]",
            "pg 0.1 is stale+active+undersized+degraded, acting [0]",
            "pg 0.2 is stale+active+undersized+degraded, acting [0]",
            "pg 0.3 is stale+active+undersized+degraded, acting [0]"
        ]
    }
}

HEALTH_WARN 2 osds down; 1 host (2 osds) down; 1 root (2 osds) down; 8 pgs stale
OSD_DOWN 2 osds down
    osd.0 (root=default,host=gnit) is down
    osd.1 (root=default,host=gnit) is down
OSD_HOST_DOWN 1 host (2 osds) down
    host gnit (root=default) (2 osds) is down
OSD_ROOT_DOWN 1 root (2 osds) down
    root default (2 osds) is down
PG_STALE 8 pgs stale
    pg 0.7 is stale+active+clean, acting [1,0]
    pg 0.6 is stale+active+clean, acting [0,1]
    pg 0.5 is stale+active+clean, acting [0,1]
    pg 0.4 is stale+active+clean, acting [0,1]
    pg 0.0 is stale+active+clean, acting [0,1]
    pg 0.1 is stale+active+clean, acting [1,0]
    pg 0.2 is stale+active+clean, acting [0,1]
    pg 0.3 is stale+active+clean, acting [0,1]
</pre>

leseb commented:
@liewegas thanks!

liewegas commented:
Okay, added a mon_health_preluminous_compat option that is off by default but can be enabled. Also moved detail into each health section; I think this is simpler.

Normal mode:
<pre>

{
    "checks": {
        "OSD_DOWN": {
            "severity": "HEALTH_WARN",
            "message": "2 osds down",
            "detail": [
                "osd.0 (root=default,host=gnit) is down",
                "osd.1 (root=default,host=gnit) is down"
            ]
        },
        "OSD_HOST_DOWN": {
            "severity": "HEALTH_WARN",
            "message": "1 host (2 osds) down",
            "detail": [
                "host gnit (root=default) (2 osds) is down"
            ]
        },
        "OSD_ROOT_DOWN": {
            "severity": "HEALTH_WARN",
            "message": "1 root (2 osds) down",
            "detail": [
                "root default (2 osds) is down"
            ]
        },
        "PG_DEGRADED": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs degraded",
            "detail": [
                "pg 0.7 is stale+active+undersized+degraded, acting [0]",
                "pg 0.6 is stale+active+undersized+degraded, acting [0]",
                "pg 0.5 is stale+active+undersized+degraded, acting [0]",
                "pg 0.4 is stale+active+undersized+degraded, acting [0]",
                "pg 0.0 is stale+active+undersized+degraded, acting [0]",
                "pg 0.1 is stale+active+undersized+degraded, acting [0]",
                "pg 0.2 is stale+active+undersized+degraded, acting [0]",
                "pg 0.3 is stale+active+undersized+degraded, acting [0]"
            ]
        },
        "PG_STALE": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs stale",
            "detail": [
                "pg 0.7 is stale+active+undersized+degraded, acting [0]",
                "pg 0.6 is stale+active+undersized+degraded, acting [0]",
                "pg 0.5 is stale+active+undersized+degraded, acting [0]",
                "pg 0.4 is stale+active+undersized+degraded, acting [0]",
                "pg 0.0 is stale+active+undersized+degraded, acting [0]",
                "pg 0.1 is stale+active+undersized+degraded, acting [0]",
                "pg 0.2 is stale+active+undersized+degraded, acting [0]",
                "pg 0.3 is stale+active+undersized+degraded, acting [0]"
            ]
        },
        "PG_UNDERSIZED": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs undersized",
            "detail": [
                "pg 0.7 is stale+active+undersized+degraded, acting [0]",
                "pg 0.6 is stale+active+undersized+degraded, acting [0]",
                "pg 0.5 is stale+active+undersized+degraded, acting [0]",
                "pg 0.4 is stale+active+undersized+degraded, acting [0]",
                "pg 0.0 is stale+active+undersized+degraded, acting [0]",
                "pg 0.1 is stale+active+undersized+degraded, acting [0]",
                "pg 0.2 is stale+active+undersized+degraded, acting [0]",
                "pg 0.3 is stale+active+undersized+degraded, acting [0]"
            ]
        }
    },
    "status": "HEALTH_WARN"
}
HEALTH_WARN 2 osds down; 1 host (2 osds) down; 1 root (2 osds) down; 8 pgs degraded; 8 pgs stale; 8 pgs undersized
OSD_DOWN 2 osds down
    osd.0 (root=default,host=gnit) is down
    osd.1 (root=default,host=gnit) is down
OSD_HOST_DOWN 1 host (2 osds) down
    host gnit (root=default) (2 osds) is down
OSD_ROOT_DOWN 1 root (2 osds) down
    root default (2 osds) is down
PG_DEGRADED 8 pgs degraded
    pg 0.7 is stale+active+undersized+degraded, acting [0]
    pg 0.6 is stale+active+undersized+degraded, acting [0]
    pg 0.5 is stale+active+undersized+degraded, acting [0]
    pg 0.4 is stale+active+undersized+degraded, acting [0]
    pg 0.0 is stale+active+undersized+degraded, acting [0]
    pg 0.1 is stale+active+undersized+degraded, acting [0]
    pg 0.2 is stale+active+undersized+degraded, acting [0]
    pg 0.3 is stale+active+undersized+degraded, acting [0]
PG_STALE 8 pgs stale
    pg 0.7 is stale+active+undersized+degraded, acting [0]
    pg 0.6 is stale+active+undersized+degraded, acting [0]
    pg 0.5 is stale+active+undersized+degraded, acting [0]
    pg 0.4 is stale+active+undersized+degraded, acting [0]
    pg 0.0 is stale+active+undersized+degraded, acting [0]
    pg 0.1 is stale+active+undersized+degraded, acting [0]
    pg 0.2 is stale+active+undersized+degraded, acting [0]
    pg 0.3 is stale+active+undersized+degraded, acting [0]
PG_UNDERSIZED 8 pgs undersized
    pg 0.7 is stale+active+undersized+degraded, acting [0]
    pg 0.6 is stale+active+undersized+degraded, acting [0]
    pg 0.5 is stale+active+undersized+degraded, acting [0]
    pg 0.4 is stale+active+undersized+degraded, acting [0]
    pg 0.0 is stale+active+undersized+degraded, acting [0]
    pg 0.1 is stale+active+undersized+degraded, acting [0]
    pg 0.2 is stale+active+undersized+degraded, acting [0]
    pg 0.3 is stale+active+undersized+degraded, acting [0]
</pre>

Compat mode:
<pre>
gnit:build (wip-health) 11:54 AM $ bin/ceph health detail -f json-pretty ; bin/ceph health detail

{
    "checks": {
        "OSD_DOWN": {
            "severity": "HEALTH_WARN",
            "message": "2 osds down",
            "detail": [
                "osd.0 (root=default,host=gnit) is down",
                "osd.1 (root=default,host=gnit) is down"
            ]
        },
        "OSD_HOST_DOWN": {
            "severity": "HEALTH_WARN",
            "message": "1 host (2 osds) down",
            "detail": [
                "host gnit (root=default) (2 osds) is down"
            ]
        },
        "OSD_ROOT_DOWN": {
            "severity": "HEALTH_WARN",
            "message": "1 root (2 osds) down",
            "detail": [
                "root default (2 osds) is down"
            ]
        },
        "PG_DEGRADED": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs degraded",
            "detail": [
                "pg 0.7 is stale+active+undersized+degraded, acting [0]",
                "pg 0.6 is stale+active+undersized+degraded, acting [0]",
                "pg 0.5 is stale+active+undersized+degraded, acting [0]",
                "pg 0.4 is stale+active+undersized+degraded, acting [0]",
                "pg 0.0 is stale+active+undersized+degraded, acting [0]",
                "pg 0.1 is stale+active+undersized+degraded, acting [0]",
                "pg 0.2 is stale+active+undersized+degraded, acting [0]",
                "pg 0.3 is stale+active+undersized+degraded, acting [0]"
            ]
        },
        "PG_STALE": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs stale",
            "detail": [
                "pg 0.7 is stale+active+undersized+degraded, acting [0]",
                "pg 0.6 is stale+active+undersized+degraded, acting [0]",
                "pg 0.5 is stale+active+undersized+degraded, acting [0]",
                "pg 0.4 is stale+active+undersized+degraded, acting [0]",
                "pg 0.0 is stale+active+undersized+degraded, acting [0]",
                "pg 0.1 is stale+active+undersized+degraded, acting [0]",
                "pg 0.2 is stale+active+undersized+degraded, acting [0]",
                "pg 0.3 is stale+active+undersized+degraded, acting [0]"
            ]
        },
        "PG_UNDERSIZED": {
            "severity": "HEALTH_WARN",
            "message": "8 pgs undersized",
            "detail": [
                "pg 0.7 is stale+active+undersized+degraded, acting [0]",
                "pg 0.6 is stale+active+undersized+degraded, acting [0]",
                "pg 0.5 is stale+active+undersized+degraded, acting [0]",
                "pg 0.4 is stale+active+undersized+degraded, acting [0]",
                "pg 0.0 is stale+active+undersized+degraded, acting [0]",
                "pg 0.1 is stale+active+undersized+degraded, acting [0]",
                "pg 0.2 is stale+active+undersized+degraded, acting [0]",
                "pg 0.3 is stale+active+undersized+degraded, acting [0]"
            ]
        }
    },
    "status": "HEALTH_WARN",
    "summary": [
        {
            "severity": "HEALTH_WARN",
            "summary": "2 osds down"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "1 host (2 osds) down"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "1 root (2 osds) down"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "8 pgs degraded"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "8 pgs stale"
        },
        {
            "severity": "HEALTH_WARN",
            "summary": "8 pgs undersized"
        }
    ],
    "overall_status": "HEALTH_WARN",
    "detail": [
        "osd.0 (root=default,host=gnit) is down",
        "osd.1 (root=default,host=gnit) is down",
        "host gnit (root=default) (2 osds) is down",
        "root default (2 osds) is down",
        "pg 0.7 is stale+active+undersized+degraded, acting [0]",
        "pg 0.6 is stale+active+undersized+degraded, acting [0]",
        "pg 0.5 is stale+active+undersized+degraded, acting [0]",
        "pg 0.4 is stale+active+undersized+degraded, acting [0]",
        "pg 0.0 is stale+active+undersized+degraded, acting [0]",
        "pg 0.1 is stale+active+undersized+degraded, acting [0]",
        "pg 0.2 is stale+active+undersized+degraded, acting [0]",
        "pg 0.3 is stale+active+undersized+degraded, acting [0]",
        "pg 0.7 is stale+active+undersized+degraded, acting [0]",
        "pg 0.6 is stale+active+undersized+degraded, acting [0]",
        "pg 0.5 is stale+active+undersized+degraded, acting [0]",
        "pg 0.4 is stale+active+undersized+degraded, acting [0]",
        "pg 0.0 is stale+active+undersized+degraded, acting [0]",
        "pg 0.1 is stale+active+undersized+degraded, acting [0]",
        "pg 0.2 is stale+active+undersized+degraded, acting [0]",
        "pg 0.3 is stale+active+undersized+degraded, acting [0]",
        "pg 0.7 is stale+active+undersized+degraded, acting [0]",
        "pg 0.6 is stale+active+undersized+degraded, acting [0]",
        "pg 0.5 is stale+active+undersized+degraded, acting [0]",
        "pg 0.4 is stale+active+undersized+degraded, acting [0]",
        "pg 0.0 is stale+active+undersized+degraded, acting [0]",
        "pg 0.1 is stale+active+undersized+degraded, acting [0]",
        "pg 0.2 is stale+active+undersized+degraded, acting [0]",
        "pg 0.3 is stale+active+undersized+degraded, acting [0]"
    ]
}
HEALTH_WARN 2 osds down; 1 host (2 osds) down; 1 root (2 osds) down; 8 pgs degraded; 8 pgs stale; 8 pgs undersized
osd.0 (root=default,host=gnit) is down
osd.1 (root=default,host=gnit) is down
host gnit (root=default) (2 osds) is down
root default (2 osds) is down
pg 0.7 is stale+active+undersized+degraded, acting [0]
pg 0.6 is stale+active+undersized+degraded, acting [0]
pg 0.5 is stale+active+undersized+degraded, acting [0]
pg 0.4 is stale+active+undersized+degraded, acting [0]
pg 0.0 is stale+active+undersized+degraded, acting [0]
pg 0.1 is stale+active+undersized+degraded, acting [0]
pg 0.2 is stale+active+undersized+degraded, acting [0]
pg 0.3 is stale+active+undersized+degraded, acting [0]
pg 0.7 is stale+active+undersized+degraded, acting [0]
pg 0.6 is stale+active+undersized+degraded, acting [0]
pg 0.5 is stale+active+undersized+degraded, acting [0]
pg 0.4 is stale+active+undersized+degraded, acting [0]
pg 0.0 is stale+active+undersized+degraded, acting [0]
pg 0.1 is stale+active+undersized+degraded, acting [0]
pg 0.2 is stale+active+undersized+degraded, acting [0]
pg 0.3 is stale+active+undersized+degraded, acting [0]
pg 0.7 is stale+active+undersized+degraded, acting [0]
pg 0.6 is stale+active+undersized+degraded, acting [0]
pg 0.5 is stale+active+undersized+degraded, acting [0]
pg 0.4 is stale+active+undersized+degraded, acting [0]
pg 0.0 is stale+active+undersized+degraded, acting [0]
pg 0.1 is stale+active+undersized+degraded, acting [0]
pg 0.2 is stale+active+undersized+degraded, acting [0]
pg 0.3 is stale+active+undersized+degraded, acting [0]
</pre>

liewegas commented:
The detail output may get some duplication and the new checks' detail messages may not be fully explanatory (they are assumed to be nested beneath a summary message), but this is at least in a form that won't break a tool.

liewegas commented:
@jcsp added MDS checks.  It was a little tricky to group the MDSHealth events since we only have a single layer of grouping overall and we have items from multiple FSs and MDSs, but I think the final result makes sense...

gregsfortytwo commented on src/mon/OSDMonitor.cc:
```diff
@@ -1382,7 +1384,365 @@ void OSDMonitor::_check_health(
+  // OSD_DOWN
+  // OSD_$subtree_DOWN
+  // OSD_ORPHAN
+  if (num_osds >= 0) {
```
This if block is a little weird following the '''goto out'''' above. Especially because it's so long and covers a bunch of different tests differently than the following blocks.

gregsfortytwo commented on src/mon/OSDMonitor.cc:
```diff
@@ -1382,7 +1384,365 @@ void OSDMonitor::_check_health(
+	continue;
+      if (!osdmap.is_up(i)) {
+	down_in_osds.insert(i);
+	int parent_id = 0;
```
Is this indentation single-space instead of 2?

gregsfortytwo commented on src/mon/PGMap.cc:
```diff
@@ -2548,6 +2550,629 @@ namespace {
+	 << roundf(max_osd_usage*1000.0)/100.0
+	 << "%) osd usage " << roundf(diff*1000.0)/100.0 << "% > "
+	 << roundf(cct->_conf->mon_warn_osd_usage_min_max_delta*1000.0)/100.0
+	 << " (mon_warn_osd_usage_min_max_delta)";
```
This should probably include a link to something explaining PG and OSD balancing.

jcsp commented on src/mon/OSDMonitor.cc:
```diff
@@ -1052,6 +1056,376 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+    }
+  }
+
+out:
```
`out` label unused

jcsp commented:
I think the 'message' field on each check should be a dict called 'summary' with a 'message' item in it.  Then, we can add more fields there with machine-readable summary info.  The items in detail should be dicts too, so it's like this:

```
"PG_STUCK_UNCLEAN": {
  "severity": "HEALTH_ERR",
  "summary": {message: "13 pgs stuck unclean", pool_ids=[4]},
  "detail": [{"message": "pg blahblah", pg_id="4.3"}]
}
```

The alternative would be to keep the "message" string field and add a separate "metadata" field for the summary metadata, but having one 'summary' object that contains both the human and machine readable toplevel info feels more natural to me.

For example, the summary part of a PG error state might be used to indicate which pools are affected (I claim that while having O(pgs) items in a summary structure is bad, having O(pools) is okay).  Ideally, callers would have enough information in the summary dict to reconstruct the message string for themselves if they want to.



The text of the log messages when health conditions come and go needs to be a lot friendlier, but we can do that in a followup.

liewegas commented:
I think that works.  One way to enforce the scale of the summary info is to make the summary string include all of the summary info.  If pools are enumerated, list them in the summary string too.

I'm not sure the best way to include this free from info in the health records.  Perhaps we can just encode JSON in a string?

jcsp commented:
In the case of listing pools, I suspect we wouldn't want to put them human-readable in the default message because the user might have picked long names.  My main motivation for wanting the list of IDs in the machine readable summary is to enable a UI to put a warning light next to pools that are associated with health issues.

I think there are a bunch of places where we're going to want to pass json-ish data around -- maybe we could do with a nice class that wraps "dict of json-ish things" into an ::encode/::decode-able form?  In thee first instance the encoding could just be JSON but it would leave us with the option to use something more efficient if we wanted.

jcsp commented on src/mon/OSDMonitor.cc:
```diff
@@ -1052,6 +1056,370 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+}
+
+void OSDMonitor::_check_health(
+  const OSDMap& nextmap,
```
All the checks in this function are meant to be using nextmap, right?  They're looking at the old map currently.

liewegas replied:
Yep.  They're also a function of OSDMap; moving to a new method there.
